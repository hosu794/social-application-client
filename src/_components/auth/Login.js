import React, { useState, useEffect } from "react";
import { Link } from "react-router-dom";
import { useDispatch, useSelector } from "react-redux";

import { authActions, alertActions } from "../../_actions";

function Login() {
  const [inputs, setInputs] = useState({
    username: "",
    password: "",
  });

  const { username, password } = inputs;
  const loading = useSelector((state) => state.authentication.loading);
  const logginIn = useSelector((state) => state.authentication.logginIn);
  const alert = useSelector((state) => state.alert);
  const dispatch = useDispatch();
  const [error, setError] = useState({
    isTrue: false,
    message: "",
  });

  function handleChange(e) {
    const { name, value } = e.target;
    setInputs((inputs) => ({
      ...inputs,
      [name]: value,
    }));
  }

  function clearAlerts() {
    setError((error) => ({
      isTrue: false,
      message: "",
    }));
  }

  function handleAlertButton() {
    clearAlerts();
  }

  useEffect(() => {
    dispatch(authActions.logout());
  }, []);

  function handleSubmit(e) {
    e.preventDefault();

    dispatch(alertActions.clear());

    isValidPasswordAndUsername();

    if (username && password) {
      dispatch(authActions.login(username, password));
      areCrudentialsAreCorrect(alert);
    }
  }

  function isValidPasswordAndUsername() {
    if (!username) {
      setError((error) => ({
        isTrue: true,
        message: "Username is required!",
      }));
    } else if (!password) {
      setError((error) => ({
        isTrue: true,
        message: "Password is required!",
      }));
    }
  }

  function areCrudentialsAreCorrect(alert) {
    if (alert.type == "alert-danger") {
      console.log("Danger");

      setError((error) => ({
        isTrue: true,
        message: "Bad Crudientials!",
      }));
    }
  }

  return (
    <section class="hero is-success is-fullheight">
      <div class="hero-body">
        <form onSubmit={handleSubmit} class="container has-text-centered">
          {error.isTrue && (
            <div class="notification is-danger">
              <button onClick={handleAlertButton} class="delete"></button>
              {error.message}
            </div>
          )}
          <div class="field">
            <p class="control has-icons-left has-icons-right">
              <input
                class="input"
                name="username"
                onChange={handleChange}
                value={username}
                type="text"
                placeholder="Username"
              />

              <span class="icon is-small is-left">
                <i class="fas fa-envelope"></i>
              </span>
            </p>
          </div>
          <div class="field">
            <p class="control has-icons-left">
              <input
                class="input"
                name="password"
                onChange={handleChange}
                value={password}
                type="password"
                placeholder="Password"
              />
              <span class="icon is-small is-left">
                <i class="fas fa-lock"></i>
              </span>
            </p>
          </div>
          <div className="field">
            <button
              type="submit"
              class="button is-primary is-fullwidth is-light"
            >
              Login
            </button>
          </div>
        </form>
      </div>
    </section>
  );
}

export default Login;
