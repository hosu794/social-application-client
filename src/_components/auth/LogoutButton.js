import React from "react";

import { useDispatch, useSelector } from "react-redux";
import { authActions } from "../../_actions";

function LogoutButton() {
  const dispatch = useDispatch();

  const user = useSelector((state) => state.user.user);

  function logoutUserOnClick() {
    dispatch(authActions.logout());
  }

  return (
    <a onClick={logoutUserOnClick} className="button is-light">
      Log out
    </a>
  );
}

export default LogoutButton;
