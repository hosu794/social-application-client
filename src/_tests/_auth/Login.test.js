import React from "react";

import configureStore from "redux-mock-store";
import { MemoryRouter } from "react-router-dom";
import { Provider } from "react-redux";
import Login from "../../_components/auth/Login";
import { mount } from "enzyme";
import LogoutButton from "../../_components/auth/LogoutButton";
import { Form } from "formik";

const mockStore = configureStore([]);

let store = mockStore({
  authentication: {
    loading: false,
    error: true,
  },
});

describe("Test for the Login Component", () => {
  const TestingComponent = () => (
    <MemoryRouter>
      <Provider store={store}>
        <Login />
      </Provider>
    </MemoryRouter>
  );

  const wrapper = mount(<TestingComponent />);

  test("should render component correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("should component be defined", () => {
    expect(Login).toBeDefined();
  });

  test("should update an first input when it is changed", () => {
    wrapper
      .find("input")
      .at(0)
      .simulate("change", {
        persist: () => {},
        target: {
          name: "username",
          value: "exampleusername",
        },
      });

    let newValue = wrapper.find(Login).find("input").at(0).props().value;

    expect(newValue).toEqual("exampleusername");
  });

  test("should update an second input when it is changed", () => {
    wrapper
      .find("input")
      .at(1)
      .simulate("change", {
        persist: () => {},
        target: {
          name: "password",
          value: "randompassword",
        },
      });

    let newValue = wrapper.find(Login).find("input").at(1).props().value;
    expect(newValue).toEqual("randompassword");
  });

  test("should show error message, when first field is empty", () => {
    wrapper.find(".button").simulate("click");

    let element = wrapper.find(".is-danger").at(0).text();
    expect(element).toBeDefined();
  });

  test("should show error message, when second field is empty", () => {
    wrapper.find(".button").simulate("click");

    let element = wrapper.find(".is-danger").at(1).text();

    expect(element).toBeDefined();
  });
});
