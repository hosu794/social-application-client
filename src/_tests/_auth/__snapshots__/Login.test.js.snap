// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for the Login Component should render component correctly 1`] = `ReactWrapper {}`;
